src/mod/applications/mod_http_cache/prometheus_metrics.cpp

```diff
@@ -0,0 +1,125 @@
+#include <map>
+#include <string>
+#include "prometheus_metrics.h"
+
+
+class prometheus_metrics
+{
+public:
+	class auto_lock
+	{
+	public:
+		auto_lock(switch_mutex_t* mutex) : _mutex(mutex)
+		{
+			switch_mutex_lock(_mutex);
+		}
+		~auto_lock()
+		{
+			switch_mutex_unlock(_mutex);
+		}
+		switch_mutex_t* _mutex;
+	};
+	
+	prometheus_metrics(switch_loadable_module_interface_t **module_interface, switch_memory_pool_t* pool) :
+		_pool(pool),
+		_module_interface(module_interface),
+		_download_duration_total(0),
+		_download_count(0),
+		_download_fail_count(0)
+	{
+		switch_mutex_init(&_mutex, SWITCH_MUTEX_NESTED, _pool);
+	}
+	
+	~prometheus_metrics()
+	{
+		switch_mutex_destroy(_mutex);
+	}
+
+	void increment_download_duration(unsigned int duration)
+	{
+		auto_lock lock(_mutex);
+		_download_duration_total += duration;
```

Comment:
I'm not sure if this is a proper way to calculate duration and I'm guessing you will have to compute this by sum + rate to calculate the average speed. Take note that our current minimum precision is 1 minute. Can you compare your approach vs this one https://github.com/team-telnyx/mod_xml_http/blob/master/prometheus_metrics.cpp#L129?

My worry is for example if we have intermittent network issue and you get this metric like in 30 seconds you got 10 request having 3000ms+ to download an audio file but in the next 30 seconds you got 40 request having 100ms to download an audio file. The 1 min would calculate around just 500ms I think and for this approach we might have miss that spike of network issue.

Reply:
Hm, agree.

But then, how about this? Rather than use `report_algorithmic_delay`, I'd like to introduce the small bucket counters.

Because we don't interesting in actual download duration, but we want to know how many times over duration was happening. And the average calculation makes it hard to find the actual problem case easily.

So I'd like to introduce the below counters something like a bucket.

Counter 1: 0~299ms 
Counter 2: 300 ~ 499 ms
Counter 3: 500 ~ 799 ms
Counter 4: 800 ~ 1299 ms
Counter 5: 1300 ~ 2099 ms
Counter 6: 2100 ~ 3399 ms
Counter 7: 3400 ~ 5499 ms
Counter 8: 5500 ~ inf

Reply:
Sure, I think that is a better way to solve it.

Reply:
Thanks, I've fixed it. :)